src/sliding-sync.ts

```diff
@@ -615,11 +622,52 @@ export class SlidingSync extends TypedEventEmitter<SlidingSyncEvent, SlidingSync
     }
 
     /**
-     * Resend a Sliding Sync request. Used when something has changed in the request.
+     * Resend a Sliding Sync request. Used when something has changed in the request. Resolves with
+     * the transaction ID of this request on success. Rejects with the transaction ID of this request
+     * on failure.
      */
-    public resend(): void {
+    public resend(): Promise<string> {
         this.needsResend = true;
+        this.txnId = ""+Math.random();
+        const p: Promise<string> = new Promise((resolve, reject) => {
+            this.txnIdDefers.push({
+                txnId: this.txnId,
+                resolve: resolve,
+                reject: reject,
+            });
+        });
         this.pendingReq?.abort();
+        return p;
+    }
+
+    private resolveTransactionDefers(txnId?: string) {
+        if (!txnId) {
+            return;
+        }
+        // find the matching index
+        let txnIndex = -1;
+        for (let i = 0; i < this.txnIdDefers.length; i++) {
+            if (this.txnIdDefers[i].txnId === txnId) {
+                txnIndex = i;
+                break;
+            }
+        }
+        if (txnIndex === -1) {
+            // this shouldn't happen; we shouldn't be seeing txn_ids for things we don't know about,
+            // whine about it.
+            logger.warn(`resolveTransactionDefers: seen ${txnId} but it isn't a pending txn, ignoring.`);
+            return;
+        }
+        // This list is sorted in time, so if the input txnId ACKs in the middle of this array,
+        // then everything before it that hasn't been ACKed yet never will and we should reject them.
+        for (let i = 0; i < txnIndex; i++) {
+            if (i < txnIndex) {
```

Comment:
Why do we need to check this again?

Reply:
If you send 5 requests immediately, then there will be 5 elements in `this.txnIdDefers`, sorted in time so `[0]` is the earliest. Some of these requests may have made it to the server, some may not have. If for example, the 3rd request made it, then we want to reject the defers for requests 1 and 2 as they were lost (this relies on a single in-flight request at a time), and resolve the defer for request 3. This conditional is doing that.

Reply:
In practice this can happen very easily when users move around the app - each time they change the UI (space, room, etc) we cancel the current long-poll, modify the req params and send a new request. Rapidly switching rooms will cause many consecutive request -> cancel operations, and we don't know if these cancelled requests made it to the server.

Reply:
I was thinking much simpler: the loop condition is `i < txnIndex` so the the if statement just does the same test as the loop condition has just done each iteration of the loop? The loop only runs for i = 0, 1 ... txnIndex anyway.

Reply:
Derp, thanks!